core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAnalyzer.scala

```diff
@@ -320,6 +322,93 @@ class AppSparkMetricsAnalyzer(app: AppBase) extends AppAnalysisBase(app) {
+        sm.stageInfo.attemptNumber())
+      // count duplicate task attempts
+      val numAttempts = tasksInStage.size
+      val nodeNames = sqlAnalyzer.stageToNodeNames.
+        getOrElse(sm.stageInfo.stageId, Seq.empty[String])
+      val disgnosticMetrics = sqlAnalyzer.stageToDiagnosticMetrics.
+        getOrElse(sm.stageInfo.stageId, mutable.HashMap.empty[String, StatisticsMetrics])
+      val zeroStats = StatisticsMetrics.ZERO_RECORD
+      val memorySpilledMetrics = disgnosticMetrics.getOrElse(MEMORY_SPILLED_METRIC, zeroStats)
+      val diskSpilledMetrics = disgnosticMetrics.getOrElse(DISK_SPILLED_METRIC, zeroStats)
+      val inputBytesReadMetrics = disgnosticMetrics.getOrElse(INPUT_BYTES_READ_METRIC, zeroStats)
+      val outputBytesMetrics = disgnosticMetrics.getOrElse(OUTPUT_BYTES_WRITTEN_METRIC, zeroStats)
+      val swTotalBytesMetrics = disgnosticMetrics.getOrElse(SW_TOTAL_BYTES_METRIC, zeroStats)
+      val srFetchWaitTimeMetrics = disgnosticMetrics.getOrElse(SR_FETCH_WAIT_TIME_METRIC, zeroStats)
+      val swWriteTimeMetrics = disgnosticMetrics.getOrElse(SW_WRITE_TIME_METRIC, zeroStats)
+      val gpuSemaphoreMetrics = disgnosticMetrics.getOrElse(GPU_SEMAPHORE_WAIT_METRIC, zeroStats)
+      val srTotalBytesMetrics =
+        AppSparkMetricsAnalyzer.getStatistics(tasksInStage.map(_.sr_totalBytesRead))
+
+      StageDiagnosticResult(index,
+        app.getAppName,
+        app.appId,
+        sm.stageInfo.stageId,
+        sm.duration,
+        numAttempts,  // TODO: why is this numAttempts and not numTasks?
+        bytesToMB(memorySpilledMetrics.min),
+        bytesToMB(memorySpilledMetrics.med),
+        bytesToMB(memorySpilledMetrics.max),
+        bytesToMB(memorySpilledMetrics.total),
+        bytesToMB(diskSpilledMetrics.min),
+        bytesToMB(diskSpilledMetrics.med),
+        bytesToMB(diskSpilledMetrics.max),
+        bytesToMB(diskSpilledMetrics.total),
+        inputBytesReadMetrics.min,
+        inputBytesReadMetrics.med,
+        inputBytesReadMetrics.max,
+        inputBytesReadMetrics.total,
+        outputBytesMetrics.min,
+        outputBytesMetrics.med,
+        outputBytesMetrics.max,
+        outputBytesMetrics.total,
+        srTotalBytesMetrics.min,
+        srTotalBytesMetrics.med,
+        srTotalBytesMetrics.max,
+        srTotalBytesMetrics.total,
+        swTotalBytesMetrics.min,
+        swTotalBytesMetrics.med,
+        swTotalBytesMetrics.max,
+        swTotalBytesMetrics.total,
+        nanoToMilliSec(srFetchWaitTimeMetrics.min),
+        nanoToMilliSec(srFetchWaitTimeMetrics.med),
+        nanoToMilliSec(srFetchWaitTimeMetrics.max),
+        nanoToMilliSec(srFetchWaitTimeMetrics.total),
+        nanoToMilliSec(swWriteTimeMetrics.min),
+        nanoToMilliSec(swWriteTimeMetrics.med),
+        nanoToMilliSec(swWriteTimeMetrics.max),
+        nanoToMilliSec(swWriteTimeMetrics.total),
+        gpuSemaphoreMetrics.total,
+        nodeNames)
```

Comment:
We have the same problem with `AppSparkMetricsAnalyzer. diagnosticMetricsMap()`
I don't fully understand why we are doing those steps in the code:

1. Inside `aggregateDiagnosticSparkMetricsByStage` we start by calling `diagnosticsMetricsMap()` which creates this transient state converting the `AccumProfileResults` to `StatisticsMetrics`. This implies that we create new objects that contain the same fields that already exist in  AccumProfileResults (min, median, max, total)
2. we aggregate the (min,median,max,total) once more in to a large case class `StageDiagnosticResult`

So the question is: why did we allocate all those `StatisticsMetrics` just to abandon them once we get outside of `diagnosticMetricsMap()`? We could achieve the same by iterating on `AccumProfileResults()` (which saves the unecessary allocation of StatisticsMetrics)

Reply:
Thanks @amahussein! Updated PR based on review feedback:

1. Removed diagnostic results from `ApplicationSummaryInfo`. Created class `DiagnosticSummaryInfo` to store diagnostic results. Now the diagnostic results will be saved to CSV file `stage_level_diagnostic_metrics.csv` only, not in the `profile.log`
2. Removed redundant function `diagnosticsMetricsMap` from `aggregateDiagnosticSparkMetricsByStage`
3. Refactored `stageToDiagnosticMetrics` to store copies of `AccumProfileResults` (memory enhancement)
4. Refactored `StageDiagnosticResult` to store copies of `AccumProfileResults` (memory enhancement)
5. Updated unit tests accordingly

Reply:
I have a question about the removal of `analysis.stageDiagnostics` from `ApplicationSummaryInfo`. After removal, we are encapsulating `analysis.stageDiagnostics` in `DiagnosticSummaryInfo`.  Finally, we access `diagnostics.stageDiagnostics` when generating the report.

Could you clarify the need for the `DiagnosticSummaryInfo` class? If  it is a wrapper for `analysis.stageDiagnostics`, we could put it in `ApplicationSummaryInfo` directly.

Reply:
> I have a question about the removal of `analysis.stageDiagnostics` from `ApplicationSummaryInfo`. After removal, we are encapsulating `analysis.stageDiagnostics` in `DiagnosticSummaryInfo`. Finally, we access `diagnostics.stageDiagnostics` when generating the report.
> 
> Could you clarify the need for the `DiagnosticSummaryInfo` class? If it is a wrapper for `analysis.stageDiagnostics`, we could put it in `ApplicationSummaryInfo` directly.

IIUC, `DiagnosticSummaryInfo` is not a field of `ApplicationSummaryInfo`
@parthosa I believe you misread the code because ApplicationSummaryInfo has so many fields that you might not notice that the last 2 values are not part of its constructors.

Reply:
Aah. My bad. Thanks for clarification.